Allow Reloading of Custom Permissions

willies952002 opened the pull request:
Add the ability to reload the custom permissions file by doing "/reload permissions"
Implements Feature Request: https://github.com/PaperMC/Paper/issues/49

Sorry about reposting/recreating this, but I was unable to re-open the original PR.

kashike commented on Spigot-Server-Patches/0085-Allow-Reloading-of-Custom-Permissions.patch:
```diff
@@ -0,0 +1,38 @@
+                 pluginManager.addPermission(perm);
+             } catch (IllegalArgumentException ex) {
+-                getLogger().log(Level.SEVERE, "Permission in " + file + " was already defined", ex);
++                getLogger().log(Level.WARNING, "Permission in " + file + " was already defined"); // Paper - Just Display as Warning
```
I'd say to leave this as `SEVERE`, log messages will change and people might filter by `SEVERE`/`WARNING`/etc.

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,77 @@
+         if (!testPermission(sender)) return true;
+ 
++        // Paper Start - Reload permissions.yml (PAPER-49)
++        if ( args.length == 1 ) {
```
extra spaces

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,77 @@
+ 
++        // Paper Start - Reload permissions.yml (PAPER-49)
++        if ( args.length == 1 ) {
++            if ( args[0].equalsIgnoreCase("permissions") || args[0].equalsIgnoreCase("perms") ) {
```
extra spaces

kashike commented on Spigot-Server-Patches/0085-Allow-Reloading-of-Custom-Permissions.patch:
```diff
@@ -0,0 +1,38 @@
++
++    // Paper start
++    @Override
++    public void reloadPermissions() {
```
This won't work. This needs to be moved to `SimplePluginManager` as:

``` java
   // Paper start
    public void clearPermissions() {
        synchronized (this) {
            this.permissions.clear();
            this.defaultPerms.get(true).clear();
            this.defaultPerms.get(false).clear();
        }
    }
    // Paper end
```

or similar.

and then:

``` java
    @Override
    public void reloadPermissions() {
        ((SimplePluginManager) this.pluginManager).clearPermissions();
        this.loadCustomPermissions();
    }
```

or similar.

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,106 @@
+     public java.util.List<String> tabComplete(CommandSender sender, String alias, String[] args) throws IllegalArgumentException
+     {
+-        return java.util.Collections.emptyList();
++        return Arrays.asList("perms"); // Paper
```
`return java.util.Collections.singletonList("permissions");`

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,106 @@
+         super(name);
+         this.description = "Reloads the server configuration and plugins";
+-        this.usageMessage = "/reload";
++        this.usageMessage = "/reload [perms|permissions]"; // Paper
```
I'd say just choose one, preferably "permissions".

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,106 @@
++
+         if (!testPermission(sender)) return true;
+ 
++        // Paper Start - Reload permissions.yml (PAPER-49)
```
Start -> start, to be consistent

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,106 @@
++                return true;
++            }
++        }
++        // Paper End
```
End -> end, to be consistent

willies952002 committed:
Allow Reloading of Custom Permissions

Add the ability to reload the custom permissions file by doing "/reload permissions"

Implements Feature Request: https://github.com/PaperMC/Paper/issues/49

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,113 @@
+     public java.util.List<String> tabComplete(CommandSender sender, String alias, String[] args) throws IllegalArgumentException
+     {
+-        return java.util.Collections.emptyList();
++        return Collections.singletonList("permissions"); // Paper
```
Don't import it.

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,113 @@
+@@ -1,25 +1,40 @@
+ package org.bukkit.command.defaults;
+ 
++import java.util.ArrayList;
```
???

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,113 @@
+ import org.bukkit.ChatColor;
+ import org.bukkit.command.Command;
+ import org.bukkit.command.CommandSender;
++import org.bukkit.plugin.Plugin;
```
???

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,113 @@
++import java.util.ArrayList;
+ import java.util.Arrays;
++import java.util.Collections;
++import java.util.List;
```
???

kashike commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,113 @@
+ 
++import java.util.ArrayList;
+ import java.util.Arrays;
++import java.util.Collections;
```
???

Ichbinjoe replied:
Is the synchronized block necessary? Can't we just catch bad calls with a

## sync catcher?

Joseph Hirschfeld
joe@ibj.io
Ichbinjoe
Skype: ichbinjoehirschfeld

aikar commented on Spigot-API-Patches/0026-Add-command-to-reload-permissions.yml.patch:
```diff
@@ -0,0 +1,99 @@
+     }
++
++    // Paper start
++    public static void reloadPermissions() {
```
I don't think this one is needed, it's not going to be a common thing, so lets not litter up the Bukkit util class.

Bukkit.getServer().reloadPermissions() is sufficient API.

willies952002 replied:
Removed
